perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/modules/ElixirBonaFideStatus.java

```diff
@@ -0,0 +1,141 @@
+
+	private static final String A_U_D_userBonaFideStatus = AttributesManager.NS_USER_ATTR_DEF + ':' + USER_BONA_FIDE_STATUS_ATTR_NAME;
+	private static final String A_U_D_userBonaFideStatusRems = AttributesManager.NS_USER_ATTR_DEF + ':' + USER_BONA_FIDE_STATUS_REMS_ATTR_NAME;
+	private static final String A_U_D_userEduPersonScopedAffiliations = AttributesManager.NS_USER_ATTR_VIRT + ':' + USER_AFFILIATIONS_ATTR_NAME;
+	private static final String A_G_D_groupAttestation = AttributesManager.NS_GROUP_ATTR_DEF + ':' + GROUP_ATESTATION_ATTR_NAME;
+
+	private RegistrarManager registrar;
+
+
+	@Override
+	public void setRegistrar(RegistrarManager registrar) {
+		this.registrar = registrar;
+	}
+
+	@Override
+	public List<ApplicationFormItemData> createApplication(PerunSession user, Application application, List<ApplicationFormItemData> data) throws PerunException {
+		return data;
+	}
+
+	/**
+	 * Add new bonaFideStatus to the user attribute.
+	 */
+	@Override
+	public Application approveApplication(PerunSession session, Application app) throws PerunException {
+		User user = app.getUser();
+		Group group = app.getGroup();
+
+		AttributesManager am = session.getPerun().getAttributesManager();
+		Attribute attestation = am.getAttribute(session, group, A_G_D_groupAttestation);
+
+		String newValue = attestation.valueAsString();
+
+		Attribute bonaFideStatus = am.getAttribute(session, user, A_U_D_userBonaFideStatus);
+
+		List<String> value = new ArrayList<>();
+		if (bonaFideStatus.getValue() == null) {
+			value = bonaFideStatus.valueAsList();
+		}
+
+		value.add(newValue);
+
+		bonaFideStatus.setValue(value);
+		am.setAttribute(session, user, bonaFideStatus);
+
+		return app;
+	}
+
+	@Override
+	public Application rejectApplication(PerunSession session, Application app, String reason) throws PerunException {
+		return app;
+	}
+
+	@Override
+	public Application beforeApprove(PerunSession session, Application app) throws PerunException {
+		User user = app.getUser();
+		Group group = app.getGroup();
+
+		AttributesManager am = session.getPerun().getAttributesManager();
+		Attribute attestation = am.getAttribute(session, group, A_G_D_groupAttestation);
+
+		if (attestation == null) {
+			throw new CantBeApprovedException("Application cannot be approved: Group does not have attestation attribute set.");
+		}
+
+		String newValue = attestation.valueAsString();
+		if (newValue == null || newValue.isEmpty()) {
+			throw new CantBeApprovedException("Application cannot be approved: Group does not have attestation value set.");
+		}
+
+		return app;
+	}
+
+	@Override
+	public void canBeApproved(PerunSession session, Application app) throws PerunException {
+
+	}
+
+	/**
+	 * Validate if the user meets criteria for applying to group.
+	 */
+	@Override
+	public void canBeSubmitted(PerunSession session, Map<String, String> params) throws PerunException {
+		User user = session.getPerunPrincipal().getUser();
```

Comment:
Same as ``beforeApprove()``, but now with ``User`` object. It might not be present, if module is used on VO. There is always a way to wrap all implementation and catch any exception and translate it to generic ``CantBeSubmittedException`` so user will see nicer message and not report error box.